src/audio/volume/volume.c

```diff
@@ -314,39 +316,41 @@ static inline int volume_set_chan(struct comp_dev *dev, int chan,
 	/* Check ramp type */
 	switch (pga->ramp) {
 	case SOF_VOLUME_LINEAR:
-
-		/* Assume the ramp length (initial_ramp [ms]) describes
-		 * time of mute to 0 dB ramp. The actual volume scale min/max
-		 * is not known. If initial_ramp is zero (or negative) use
-		 * step size that reaches 0 dB gain in one step (VOL_ZERO_DB).
-		 */
-		if (pga->initial_ramp > 0)
-			inc = MAX(VOL_RAMP_STEP_CONST / pga->initial_ramp, 1);
-		else
-			inc = VOL_ZERO_DB;
-
-		/* Scale the increment with actual volume range if it was
-		 * passed via IPC.
-		 */
-		if (cd->vol_ramp_range)
-			inc = q_multsr_32x32(inc, cd->vol_ramp_range,
-					     Q_SHIFT_BITS_32(VOL_QXY_Y,
-							     VOL_QXY_Y,
-							     VOL_QXY_Y));
-
+		/* Get volume transition delta and absolute value */
 		delta = cd->tvolume[chan] - cd->volume[chan];
 		delta_abs = ABS(delta);
 
-		/* If the ramp completion would take longer than initial_ramp
-		 * increase the step size to make this ramp complete at
-		 * initial_ramp time. Calculate estimated ramp length [us].
+		/* The ramp length (initial_ramp [ms]) describes time of mute
+		 * to vol_max unmuting. Normally the volume ramp has a
+		 * constant linear slope defined this way and variable
+		 * completion time. However in streaming start it is feasible
+		 * to apply the entire topology defined ramp time to unmute to
+		 * any used volume. In this case the ramp rate is not constant.
+		 * Note also the legacy mode without known vol_ramp_range where
+		 * the volume transition always uses the topology defined time.
 		 */
-		t_us = delta_abs * VOL_RAMP_UPDATE_US / inc;
-		if (t_us > pga->initial_ramp * 1000) {
-			inc = (int32_t)((int64_t)inc * delta_abs / VOL_ZERO_DB);
-			inc = MAX(inc, 1);
+		if (pga->initial_ramp > 0) {
+			if (constant_rate_ramp && cd->vol_ramp_range > 0)
+				inc = q_multsr_32x32(cd->vol_ramp_range,
+						     VOL_RAMP_STEP_CONST,
+						     Q_MUL_SHIFT);
+			else
+				inc = q_multsr_32x32(delta_abs,
+						     VOL_RAMP_STEP_CONST,
+						     Q_MUL_SHIFT);
+
+			/* Divide and round to nearest. Note that there will
+			 * be some accumulated error in ramp time the longer
+			 * the ramp and the smaller the transition is.
+			 */
+			inc = (2 * inc / pga->initial_ramp + 1) >> 1;
+		} else {
+			inc = delta_abs;
```

Comment:
@singalsu The else part seems to be simplified as well.  But yeah, I guess this is on purpose as well. It would seem "inc = delta_abs" is correct.

Reply:
Yes the behavior is changed a bit with ramp time (length) set to zero. Earlier version used mute to 0 dB increment as step and with high gains the ramp length could become few milliseconds. It's a small mistake in previous version that I realized only now after your comment.